tests/memory_tests.cpp

```diff
@@ -181,14 +169,12 @@ TYPED_TEST(MemoryManagerTest, AllocateTB) {
     }
     else {
         ASSERT_FAILURE( RMM_ALLOC(&a, size_tb, stream) );
-        ASSERT_FAILURE( RMM_FREE(a, stream) );
```

Comment:
I think there are potential memory leaks in these tests. What if the allocation succeeds? It won't be freed.

How about we change all of these to expect failure on the alloc, and then call RMM_FREE with no expectation?

Reply:
Good point. I restored the `RMM_FREE` calls, minus the `ASSERT`s.

Reply:
Actually, it's not possible for the allocation to succeed because `ASSERT_FAILURE` will trigger if the allocation succeeds, in which case the process will be killed and the memory will be free'd anyways.

Furthermore, now we're in UB territory because we're passing an invalid pointer to `RMM_FREE`.

Reply:
Does test ASSERT kill the failure or just the test? If another test runs with the same fixture will it be in the same process even after a prior test failure?

Reply:
Ah, so it only kills the current test:
>ASSERT_* versions generate fatal failures when they fail, and abort the current function.

https://github.com/google/googletest/blob/master/googletest/docs/primer.md

But in that case, if the `RMM_ALLOC` succeeds (`ASSERT_FAILURE` fails...), then the `RMM_FREE` will never be called and there is still a memory leak until the test process exits.

Reply:
We could change `ASSERT_` to `EXPECTS_`, but then we run back into the problem of calling `RMM_FREE` on an invalid pointer when `RMM_ALLOC` fails as expected in the happy path. So we'd have to do something like:

```
auto status = RMM_ALLOC(&ptr,...);
EXPECT_EQ(status, RMM_FAILURE);
if(status == RMM_SUCCESS)
   RMM_FREE(ptr);
```

Reply:
It won't hurt to RMM_FREE an invalid pointer -- just don't check the result. Right?

Reply:
As we move towards `RMM_FREE` not returning anything, (i.e., `void RMM_FREE(...)`), we'll need to document what happens if you pass in an invalid pointer. I'd vote for taking the approach of `std::free` and say that it is UB:

>[The behavior is undefined if the value of ptr does not equal a value returned earlier by std::malloc, std::calloc, std::aligned_alloc (since C++17), or std::realloc.](https://en.cppreference.com/w/cpp/memory/c/free)

However, it also says:
 
> The function accepts (and does nothing with) the null pointer to reduce the amount of special-casing. Whether allocation succeeds or not, the pointer returned by an allocation function can be passed to std::free.

So this test would be a case where the `ptr` does result from a failed allocation call, which should be well-defined to pass it to a free. In which case, we'll have to make sure that the resources all set the `ptr` to `nullptr` in `do_allocate` when the allocation fails.

Reply:
Looks like `rmm::alloc` already takes care of this: https://github.com/rapidsai/rmm/blob/branch-0.11/include/rmm/rmm.hpp#L140

Reply:
That was a fun dive into minutiae. This PR should be good to go now.

Reply:
Yep, failed alloc should always return a nullptr, just like malloc.